### ISSUE TITLE
[proposal] Vendor Neutral CBRS Domain Proxy

### ISSUE THREAD

**Issue by jpad-freedomfi** [magma/magma#5514]
In our CBRS deployments we have seen scenarios that warrant building a vendor neutral CBRS domain proxy:

1. Instability of electricity and/or backhaul at enterprise cell sites
2. Instability of SAS - CBSD interface when connection is unreliable (e.g. #1) requiring manual intervention to remedy
3. Mixed vendor environments where channel planning becomes complicated or inflexible
4. Variability in CBSD SAS Client behavior, including brittleness

The proposal is to build a vendor neutral domain proxy. The SAS facing interface will use the WInnForum standardized SAS-CBSD interface. The CBSD/DP side will support the WInnForum SAS-CBSD interface as well as a standardized profile for TR-069, SNMP, and potentially other protocols.

A requirements / design document will follow.

**Comment by electronjoe** [comment]
NIce, I was hoping you were able to add `type: proposal` label.

As you probably see now - commenting with this GH Issue number in your PR causes it to be linked here for traceability. You can alternatively put the issue # in your PR description (instead of comment), or you can put a comment here on the GH issue pointing at the PR # and that will drop notifications over in the PR.  Any of these is fine, though perhaps I should update documentation to ask that we put GH Issue # in the PR description just so it's normalized?

**Comment by karthiksubraveti** [comment]
Thanks @119Vik for sharing the new proposal. Posting some questions and feedback here (https://github.com/119Vik/magma/blob/dp_design_%232/docs/readmes/proposals/p010_vendor_neutral_dp.md)

![Screen Shot 2021-04-22 at 6 25 25 AM](https://user-images.githubusercontent.com/8224854/115722118-8847d280-a333-11eb-97c1-bc0bdcdda128.png)
May be my ignorance here, but just wondering if load balancers support TR069 as well ?

![Screen Shot 2021-04-22 at 6 27 07 AM](https://user-images.githubusercontent.com/8224854/115722353-c47b3300-a333-11eb-99e4-092c6f0166cd.png)
Is this sequence diagram involving storing the registration request in DB relevant for phase 1(passive mode) itself or is it for later phases ? Also in one of the boxes here there is a comment mentioning "store the  new registration in DB if required", What would be the case which would not require the request to be stored in DB ? Is this the case for which the CBSD has already been registered.

- What happens when a device is replaced or decommissioned ? Is there any procedure to relinquish the registration or swap the registration with some new replaced device ?


One final comment is, I had mentioned this in previous proposal as well, Currently all services in Orc8r are in golang, It will be nice if majority of elements are in golang, I understand that microservice architecture can ensure compatibility. In reality it is simpler to have one language(golang) so that we can continue to rely on existing client libraries, tooling around tests/linters etc and reduces additional maintenance for upgrading to latest versions of language in case support for a version reaches end of life. 

In summary, I like this proposal, It is simpler and has fewer failure elements in general and will scale to the number of devices in the network rather than the requests going up constantly.

**Comment by 119Vik** [comment]
@karthiksubraveti

> May be my ignorance here, but just wondering if load balancers support TR069 as well?

Good question, I don't think there's a "TR69 LB" in the industry, however,   as per https://en.wikipedia.org/wiki/TR-069 TR69 is HTTP-based SOAP, and I think we have LBs that can manage HTTP? I assume some routing may be required at API gateway level. In fact, LBs and boxes, with X.1 and X.2 indexes should emphasize that all of the components should be HA as well as "load balancing" is preferred over "active standby".


> Is this sequence diagram involving storing the registration request in DB relevant for phase 1(passive mode) itself or is it for later phases ?

I think we may start using DB starting with phase 1 as persistent storage of "CBSD to SAS communications".  To make sure that Radio Controller microservice is stateless. 

>  Also in one of the boxes here there is a comment mentioning "store the new registration in DB if required", What would be the case which would not require the request to be stored in DB ? Is this the case for which the CBSD has already been registered.

In case CBSD is making a new request we need to store it in DB in "Pending State". In case we got a response from SAS same record will be updated to some kind of "Processed State".  In case when the response from SAS was not communicated back to CBSD(because of a network issue or PC container unexpected restart) CBSD can re-send the request and get a response from Radio Controller DB. 

> What happens when a device is replaced or decommissioned ? Is there any procedure to relinquish the registration or swap the registration with some new replaced device ?

As per CBSD to SAS communication spec I don't think we can do it. In case it's allowed by SAS we can manage it in Phase2. Seems to be a grade use-case for Phase 2, as long as it's supported by SAS OR we are allowed to do on the fly CBSD IDs translation.

> One final comment is, I had mentioned this in previous proposal as well, Currently all services in Orc8r are in golang, It will be nice if majority of elements are in golang, I understand that microservice architecture can ensure compatibility. In reality it is simpler to have one language(golang) so that we can continue to rely on existing client libraries, tooling around tests/linters etc and reduces additional maintenance for upgrading to latest versions of language in case support for a version reaches end of life.

It makes sense. From the very beginning, we considered Python and Go. We found out that other important components are written in python (enodebD and DP sertification  software), that's why we decided to move with python so far. And it seems to valid for ProtocolController and SAS Client side.

**Comment by hcgatewood** [comment]
[x-posting [my comment](https://github.com/magma/magma/pull/6246#issuecomment-825267745) here because seems like this is the relevant location]

I know we've already brought up the queue-vs-no-queue discussion, and as I said I'm on the side of no-queue as the default.

That said, I think there's another discussion that's missing -- Python vs. Go.

To me, Python categorically *should not* be used for production applications. The only way Python's superior to Go for cloud-native, production-ready deployments is developer familiarity.

However, I think it's a clear decision to go with Go here, let me explain why

1. Magma stack is Go and C++. While the Magma project is currently a polyglot project with Python, Go, C, and C++, it's widely agreed that, if we could go back in time, we would *not* write AGW in Python, as it's caused an absolute abundance of maintainability issues, CI headaches, SEVs, and performance problems. Second, the C code is also a huge pain, and we've made steady but incomplete progress moving toward C++. That all to say, the *goal* of the Magma project should be Go and C++ -- this set is minimal, production-grade, and modern.

2. Cloud stack is Go. The cloud-native community as a whole has embraced Go, and the Magma/Orc8r project in particular has comparatively deep Go expertise and a wealth of existing, established code and constructs. Adding a new language to the cloud stack may be productive in the short term, but over the long term it will cause pain and headache having to maintain a multi-language, multi-stack cloud platform.

For a more systematized, and fun read, advocating *against* polyglot projects, please do skim the classic "Choose Boring Technology", starting at slide 31 http://boringtechnology.club/#31.

I get that FF is acquainted with Python. But to me this is short-term thinking. We want DP to be a long-term contribution to the open source community, and it will need to be improved and maintained past its initial implementation. Long-term thinking says: how can we minimize the effort needed to maintain this sub-project over time? To me, Go fits as a better answer to this question. And I'm 100% confident the existing Magma Go developers will be available to provide expertise, code reviews, and design chats.

Finally, Go is a simple language, and straightforward to pick up. 90% of what you need to know is in https://golang.org/doc/effective_go, and with a (slow) reading speed of 200 words/minute, this will take 90 minutes to get through. I view that investment as worth the long-term homogeneity of the Magma cloud platform.

Let me know your thoughts. cc @119Vik @jpad-freedomfi @xbend

**Comment by xbend** [comment]
@electronjoe @karthiksubraveti @hcgatewood 
I have updated the original proposal document in (https://github.com/magma/magma/pull/6246/)
I have added:
* Application requirement estimates and database use estimates
* Updated the proposal to refrect the discussion about queue/no-queue discussion and the outcome, which will be to go with the no-queue approach in favor of pure DP usage
* Reused the data proposed by @119Vik in (https://github.com/magma/magma/pull/6298) with small modifications:
  * split the responsibility for request handling
  * RC will be storing requests for processing and check for statuses
  * CC will poll for requests and update the statuses
  * CC will implement the SAS client internally without exposing it as a separate service

**Comment by karthiksubraveti** [comment]
@xbend @119Vik @jpad-freedomfi 

Thanks for the updated proposal. LGTM. I have few minor comments below. 
```
Configuration Controller
CC fetches pending requests from the DB, batches them and sends them to SAS.
After receiving the response from SAS, it updates the DB state for the requests processed.
```
I like the idea of adding configuration controller. It abstracts the details from SAS client and radio controller and gives a really good way to add more features and customize this in the future. (I can see some CBSD specific features easily enabled through CC, perhaps hypothetical such as selecting different SAS servers for a set of CBSDs, reject or enabling select CBSD, pre allocating spectrum through some sort of a virtual CBSD etc)
One minor nit in the the way I would look at this is, CC would fetch the CBSD state(instead of requests) from database and send requests for CBSD still in pending state. The reason why this distinction is important(imo) is looking at CBSD request pigeon holes us for this usecase, looking at state enables us to expand the possibilities which can be done through the configuration controller.

**Comment by hcgatewood** [comment]
Closing as accepted
